src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/BinaryReaderExtensions.cs

```diff
@@ -70,36 +74,67 @@ internal static object ReadPrimitiveValue(this BinaryReader reader, PrimitiveTyp
             PrimitiveType.Single => reader.ReadSingle(),
             PrimitiveType.Double => reader.ReadDouble(),
             PrimitiveType.Decimal => decimal.Parse(reader.ReadString(), CultureInfo.InvariantCulture),
-            PrimitiveType.DateTime => CreateDateTimeFromData(reader.ReadInt64()),
+            PrimitiveType.DateTime => CreateDateTimeFromData(reader.ReadUInt64()),
             _ => new TimeSpan(reader.ReadInt64()),
         };
 
-    // TODO: fix https://github.com/dotnet/runtime/issues/102826
     /// <summary>
     ///  Creates a <see cref="DateTime"/> object from raw data with validation.
     /// </summary>
-    /// <exception cref="SerializationException"><paramref name="data"/> was invalid.</exception>
-    internal static DateTime CreateDateTimeFromData(long data)
+    /// <exception cref="SerializationException"><paramref name="dateData"/> was invalid.</exception>
+    internal static DateTime CreateDateTimeFromData(ulong dateData)
     {
-        // Copied from System.Runtime.Serialization.Formatters.Binary.BinaryParser
-
-        // Use DateTime's public constructor to validate the input, but we
-        // can't return that result as it strips off the kind. To address
-        // that, store the value directly into a DateTime via an unsafe cast.
-        // See BinaryFormatterWriter.WriteDateTime for details.
+        ulong ticks = dateData & 0x3FFFFFFF_FFFFFFFFUL;
+        DateTimeKind kind = (DateTimeKind)(dateData >> 62);
 
         try
         {
-            const long TicksMask = 0x3FFFFFFFFFFFFFFF;
-            _ = new DateTime(data & TicksMask);
+            return ((uint)kind <= (uint)DateTimeKind.Local) ? new DateTime((long)ticks, kind) : CreateFromAmbiguousDst(ticks);
         }
         catch (ArgumentException ex)
         {
-            // Bad data
             throw new SerializationException(ex.Message, ex);
         }
 
-        return Unsafe.As<long, DateTime>(ref data);
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static DateTime CreateFromAmbiguousDst(ulong ticks)
+        {
+            // There's no public API to create a DateTime from an ambiguous DST, and we
+            // can't use private reflection to access undocumented .NET Framework APIs.
+            // However, the ISerializable pattern *is* a documented protocol, so we can
+            // use DateTime's serialization ctor to create a zero-tick "ambiguous" instance,
+            // then keep reusing it as the base to which we can add our tick offsets.
+
+            if (s_baseAmbiguousDstDateTime is not DateTime baseDateTime)
+            {
+#pragma warning disable SYSLIB0050 // Type or member is obsolete
+                SerializationInfo si = new(typeof(DateTime), new FormatterConverter());
+                si.AddValue("ticks", 0L); // legacy value (serialized as long) - specify both just to be safe
```

Comment:
How does specifying `tick` makes us safe?

I think it can only hide problems and produce invalid values instead of throwing an exception. I cannot think about a case where it actually helps.

Reply:
I can not speak on the behalf of @GrabYourPitchforks (who authored the code), but my understanding is that initially the `SerializationInfo` for `DateTime` contained only `ticks` field. Later `dateData` was introduced, but the runtime kept emitting the old field in case it could be deserialized with older runtime.

I've double checked the code:

https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/mscorlib/system/datetime.cs#L388-L389

https://github.com/dotnet/runtime/blob/eb655cf84d8ebe8f52c83a4fab6ef577c4a536c6/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L837-L839

And you are right, in case this code were executed on a very old runtime, we would provide an invalid result. I've removed it and added comment.